### ISSUE TITLE
Keypoint choosing.

### ISSUE THREAD

**Issue by Drow999** [benjiebob/WLDO#3]
Hi Benjamin, your WLDO is excellent, I applied it on other animals and it can estimate their pose very good. However, I have a question about the keypoint you choose. For the dogs, I only see one keypoint on the whole torso, the tail-start. If there is another point around the neck, then it seems reasonable. However, there are only few points on the head. How do this one point define the direction of the torso? Or the tail-start point and the four points (point 4,2,5,11) can define the whole body. Thank you.

**Comment by benjiebob** [comment]
Thanks for your kind words! Apologies if I am misunderstanding, but I think you are asking how WLDO is able to accurately predict the global rotation of the model with only a few annotated torso points?

This is a difficult one to answer... but it is due to a combination of the 2D keypoint losses and priors which keep shape and pose parameters constrained to plausible values. When these losses are applied over a large training dataset (see https://github.com/benjiebob/StanfordExtra), the network eventually learns to do the 'right thing'... most of the time :) 

Let me know if that answers your question!

**Comment by Drow999** [comment]
Yes, exactly. If there is no more keypoint on torso, it seems hard to define the the rotation. If everything is done in the network, then I can hardly do anything. I'm considering whether the shape and the rotation of the main body can be bascially constrainted by point 4,2,5,11, the four points on the leg.

**Comment by benjiebob** [comment]
Hmm... I'm not quite sure what you are trying to achieve? Are you training your own network somehow, running an optimizer or just using the results of WLDO? 

If you just want to know the global orientation of the predicted WLDO dog mesh, you can look at the output SMBLD parameters.

**Comment by Drow999** [comment]
by now, I just want to use the result of WLDO, if it's necessary, I will do some finetuning. 

It's sound interesting, is the output SMBLD parameters in the params.npz?

**Comment by benjiebob** [comment]
For example, running demo.py outputs *.jpg and *.npz files. The npz file contains the following keys:

```
['imgname', 'pose', 'betas', 'camera', 'trans', 'has_bbox']
```

Pose: 105 dimensions. Think of as 35 x 3 in Rodrigues axis angle. The first 1 x 3 parameters are the global orientation, the rest are for each model joint to affect the limb orientations.
Betas: 26 dimensions, 20 SMAL shape coefficients + 6 SMBLD log-scale parameters.
Translation: 3 dimensions (x,y,z)

To understand these parameters a bit more, you might find this app useful: https://github.com/benjiebob/SMALViewer

Let me know if you have trouble with these parameters :)

**Comment by Drow999** [comment]
Ok, I think I have to spend time on this. I cannot understand "the rest are for each model joint to affect the limb orientations" exactly. In the SMALViewer, you use the joints to rig each part of the body? It's hard for me to understand use only one joints(one point) to change a pose. If we want to change a pose of human's arm for example, we need to fix shoulder, and then move the joint of elbow, right? 
I think you have 20 joint in one model, why are there 34x3 paramters left.

**Comment by benjiebob** [comment]
For training and evaluation, it is necessary to create a correspondence between points on the 3D SMAL model and our manually annotated 2D keypoints in StanfordExtra. 

The SMAL model has 34 model joints which are used to rotate the limbs (arms, legs etc.) of the 3D mesh. The 34 x 3 pose parameters define these 3D rotations. [Note that SMALViewer uses an earlier version of SMAL so you *might* not see the same dimensions] 

The StanfordExtra dataset has 20 keypoint locations annotated per image. Some of these keypoints refer to SMAL model joints but not all. For example, StanfordExtra contains an annotated nose point but there isn't one defined in the original SMAL model. 

To overcome this, we add some additional 6 joints to the original SMAL model (although note we never actually used the eye points).
https://github.com/benjiebob/WLDO/blob/master/global_utils/smal_model/smal_torch.py#L188

The correspondence to StanfordExtra keypoints is then defined here: https://github.com/benjiebob/WLDO/blob/master/global_utils/config.py#L72
